pkg/shares/share_merging.go

```diff
@@ -176,5 +176,58 @@ func ParseShares(rawShares [][]byte) ([]ShareSequence, error) {
 		sequences = append(sequences, currentSequence)
 	}
 
+	for _, sequence := range sequences {
+		if err := sequence.validSequenceLength(); err != nil {
+			return sequences, err
+		}
+	}
+
 	return sequences, nil
 }
+
+// validSequenceLength extracts the sequenceLength written to the first share
+// and returns an error if the number of shares needed to store a sequence of
+// length sequenceLength doesn't match the number of shares in this share
+// sequence. Returns nil if there is no error.
+func (s ShareSequence) validSequenceLength() error {
+	if len(s.Shares) == 0 {
+		return nil
+	}
+	firstShare := s.Shares[0]
+	sequenceLength, err := firstShare.SequenceLength()
+	if err != nil {
+		return err
+	}
+
+	var numberOfSharesUsed int
+	if firstShare.isCompactShare() {
+		numberOfSharesUsed = CompactSharesUsed(int(sequenceLength))
+	} else {
+		numberOfSharesUsed = MsgSharesUsed(int(sequenceLength))
+	}
+
+	if len(s.Shares) != numberOfSharesUsed {
+		return fmt.Errorf("share sequence has %d shares but expected %d shares", len(s.Shares), numberOfSharesUsed)
+	}
+	return nil
+}
+
+// CompactSharesUsed returns the number of compact shares used to store a
+// sequence of length sequenceLength.
+func CompactSharesUsed(sequenceLength int) (sharesUsed int) {
+	if sequenceLength == 0 {
+		return 0
+	}
+
+	if sequenceLength < appconsts.FirstCompactShareContentSize {
```

Comment:
[question]
by sequence, are we referring to a sequence of shares, a sequence of compact units, or a sequence of bytes?

Reply:
In this context: `sequenceLength` is the number of bytes of message, transaction, ISR, or evidence data in a sequence of shares.

A sequence of shares is applicable to both sparse and compact shares.

Reply:
Resolved in https://github.com/celestiaorg/celestia-app/pull/865/commits/84df6ae7c8575e8c49fc9b20c52879f8a0440037